starsessions/backends/redis.py

```diff
@@ -26,13 +26,15 @@ def __init__(
             redis_key_func (typing.Callable[[str], str], optional): Customize redis key name. Defaults to None.
             expire (int, optional): Key expiry in seconds. Defaults to None.
         """
-        assert url or connection, 'Either "url" or "connection" arguments must be provided.'
```

Comment:
`assert` is fine here. This line should raise while development to signal about misconfiguration.
When you run python with optimizations, then asserts are not evaluated and this line is mostly zero-costy.
See https://docs.python.org/3/using/cmdline.html#cmdoption-O

Reply:
See also https://stackoverflow.com/questions/944592/best-practice-for-using-assert/945135#945135

Reply:
But what if url or connection are not provided in production?

Isn't it better to fail at application startup directly? 

Instead of having it seem like the application is running OK, but then when a session gets created things fail and we hope that the developer detects that failure in time and has monitoring in place.

Reply:
How is it possible? :) If you face it in prod, you have another problem then.

Reply:
Imagine you misconfigured your prod environment and forgot the REDIS_URL environment variable, so it defaulted to None (with os.getenv). Then None gets passed in as the redis url and the session backend does not error out.

I believe url / connection are an absolute requirement by the contructor, and they can't both be None. If they both are, then the class simply doesn't work. That's why I don't think this should be an assert :)

Reply:
Input/env validation should be done in app's settings during startup. That's out of scope of session library. 
For us, it is important that you set either one or another. `assert` perfectly does it.

Reply:
Yes I agree that env validation should be done by the developer. But you're assuming that developers will always have validated settings, which is a really big assumption.

But I don't agree that its out of scope of the session library. I'm not proposing we validate environment variables. We are simply validating the arguments to the constructor.
If it's a required argument, why use an assert versus raise a validation error?

This makes me think of Pydantic BaseModels. If you have a required argument that isn't provided, the constructor will raise a ValidationError. In a sense, the redis backend expects a "model" with a required argument (url or connection). But we're not raising an error if its not there.


- What will it cost `starsessions` to raise an error if the constructor gets NoneTypes for both? Nothing the way I see it.
- What would the benefit be to raise the error? A much better developer experience, even if they don't notice it. Errors on startup instead of mid-execution.

The developers who don't have validated settings will 99% not have monitoring in place, so that error is just going to break their app until hopefully someone complains or the devs notice by chance. We can help prevent that scenario by simply raising this error 😄

Reply:
Okay, almost convinced. Can you pls extract all exceptions into `exceptions.py` module and raise `SessionError` or subclass instead of just `Exception`? That would be a more specific hint.

Reply:
Done in https://github.com/alex-oleshkevich/starsessions/pull/24

Reply:
Agreed, will do :)

Reply:
Good?

Reply:
Totally